File: Dynamic programming/790_DominoTrominoTiling.cs
// 790. Domino And Tromino Tiling
// Tags: dynamic programming, math-heavy
//
// Note: I couldn't solve this one on my own. The math was difficult. Need to review.

public class Solution 
{
    public int NumTilings(int n)
    {
        if(n == 1 || n == 2)
            return n;

        const int M = 1_000_000_007;
        long result = 0;
        long runningSum = 0;
        long p1 = 2;
        long p2 = 1;

        for(int i = 3; i <= n; i++)
        {
            result = (p1 + p2 + 2*(runningSum + 1)) % M;

            runningSum = (runningSum + p2) % M;

            p2 = p1;
            p1 = result;
        }

        return (int)result;
    }
}

File: Dynamic programming/790_DominoTrominoTiling.rs
// 790. Domino And Tromino Tiling
// Tags: dynamic programming, math-heavy
//
// Note: I couldn't solve this one on my own. The math was difficult. Need to review.

const MOD: i64 = 1_000_000_007;

impl Solution {
    pub fn num_tilings(n: i32) -> i32 {
        if n == 1 || n == 2 {
            return n;
        }

        let mut result: i64 = 0;
        let mut running_sum: i64 = 0;
        let mut p1: i64 = 2;
        let mut p2: i64 = 1;

        for i in 3..=n {
            result = (p1 + p2 + 2*(running_sum + 1)) % MOD;

            running_sum = (running_sum + p2) % MOD;

            p2 = p1;
            p1 = result;
        }

        return result as i32;
    }
}